singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+    printf("%s rho0:%e T0:%e G0:%e Cv0:%e E0:%e S0:%e Pmin:%e\n"
+           "non-zero elements in K0toK40 array:\n",
+           st, _rho0, _T0, _G0, _Cv0, _E0, _S0, _Pmin);
+    for (int i = 0; i < 41; i++) {
+      if (_K0toK40[i] * _K0toK40[i] > 0.0) printf("K%i:%e\t", i, _K0toK40[i]);
+    }
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("PowerMG"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
+  Real _rho0, _T0, _G0, _Cv0, _E0, _S0, _Pmin;
+  static constexpr const int PressureCoeffsK0toK40Size = 41;
+  int _M;
+  Real _K0toK40[PressureCoeffsK0toK40Size];
+  void _CheckPowerMG();
+  void _InitializePowerMG(const Real *expcoeffs);
+  PORTABLE_INLINE_FUNCTION Real _HugPressureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _HugTemperatureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _SN2Mp2(const Real x) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityTemperature(const Real rho, const Real temp) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityInternalEnergy(const Real rho, const Real sie) const;
+};
+
+inline void PowerMG::_CheckPowerMG() {
+
+  if (_rho0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter rho0 < 0");
+  }
+  if (_T0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter T0 < 0");
+  }
+  if (_Cv0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter Cv0 < 0");
+  }
+  if (_K0toK40[0] < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter K0 < 0");
+  }
+  if (_Pmin >= 0.0) {
+    _Pmin = -1000 * _K0toK40[0];
+#ifndef NDEBUG
+    PORTABLE_WARN(
+        "PowerMG model parameter Pmin not set or positive. Reset to default (-1000*K0)");
+#endif // NDEBUG
+  }
+}
+
+inline void PowerMG::_InitializePowerMG(const Real *K0toK40input) {
+
+  // The PressureCoeffsK0toK40Size (=41) allowed K0 to K40 coefficients
+  // for the pressure reference curve vs rho
+  // are seldom all used so did not want to crowd the argument list with them.
+  // Instead I ask the host code to send me a pointer to this array so that I can
+  // copy it here. Not used coeffs should be set to 0.0 (of course).
+  for (int ind = 0; ind < PressureCoeffsK0toK40Size; ind++) {
+    _K0toK40[ind] = K0toK40input[ind];
+    if (_K0toK40[ind] != 0.0) _M = ind;
+  }
+}
+
+PORTABLE_INLINE_FUNCTION Real PowerMG::_HugPressureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
```

Comment:
Looks like `eta` is fixed for the lifetime of the function from here, so I think it's nicer and more clear to mark it `const`.
```suggestion
  const Real eta = 1.0 - robust::ratio(_rho0, rho);
```
Actually you might consider implementing a private helper function for `eta` just to write less code. Something like (in private scope)
```C++
PORTABLE_INLINE_FUNCTION Real _GetEta(Real rho) const { return 1.0 - robust::ratio(_rho0, rho); }
```
Since other quantities are updated in the loop, marking the ones that aren't modified as `const` I think helps with clarity.

Reply:
Agree.

Reply:
Done